# US Domestic Fligths for the year 2008
## by Serdar Celebi

## Preliminary Wrangling

> Data gathered from https://community.amstat.org/jointscsg-section/dataexpo/dataexpo2009 to understand the flight patterns of the US domestic airlines between 1988 and 2008.

In [35]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import bz2
import csv
import glob
import os, json
import datetime, time

%matplotlib inline

> Load in your dataset and describe its properties through the questions below.
Try and motivate your exploration goals through this section.

In [47]:
df_2008 = pd.read_csv(r"C:\Users\serda\Downloads\communicate-data-project-template\Project Template\DataExpo2009\2008.csv.bz2",compression='bz2')
df_2008

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
3        2008      1           3          4    926.0         930   1054.0   
4        2008      1           3          4   1829.0        1755   1959.0   
...       ...    ...         ...        ...      ...         ...      ...   
7009723  2008     12          13          6   1002.0         959   1204.0   
7009724  2008     12          13          6    834.0         835   1021.0   
7009725  2008     12          13          6    655.0         700    856.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0              2225            WN        335  ...    4.0      8.0          0   
1              1000            WN       3231  ...    5.0     10.0          0   
2               750            WN        448  ...    3.0     17.0          0   
3              1100            WN       1746  ...    3.0      7.0          0   
4              1925            WN       3920  ...    3.0     10.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
7009723        1150            DL       1636  ...    6.0     45.0          0   
7009724        1023            DL       1637  ...    5.0     23.0          0   
7009725         856            DL       1638  ...   24.0     12.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                     NaN         0           NaN          NaN      NaN   
1                     NaN         0           NaN          NaN      NaN   
2                     NaN         0           NaN          NaN      NaN   
3                     NaN         0           NaN          NaN      NaN   
4                     NaN         0           2.0          0.0      0.0   
...                   ...       ...           ...          ...      ...   
7009723               NaN         0           NaN          NaN      NaN   
7009724               NaN         0           NaN          NaN      NaN   
7009725               NaN         0           NaN          NaN      NaN   
7009726               NaN         0           NaN          NaN      NaN   
7009727               NaN         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
3                  NaN                NaN  
4                  0.0               32.0  
...                ...                ...  
7009723            NaN                NaN  
7009724            NaN                NaN  
7009725            NaN                NaN  
7009726            NaN                NaN  
7009727            NaN                NaN  

[7009728 rows x 29 columns]

In [37]:
df_2008.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [38]:
#to merge all files
# path = r'C:\Users\serda\Downloads\communicate-data-project-template\Project Template\DataExpo2009'
# all_files = glob.glob(path + "/*.bz2")

# li = []

# for filename in all_files:
#     df = pd.read_csv(filename, index_col=None, header=0)
#     li.append(df)

# frame = pd.concat(li, axis=0, ignore_index=True)

In [39]:
print(df_2008.shape)
print(df_2008.dtypes)

(7009728, 29)
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


In [ ]:
df_2008['DepTime']

In [40]:
datetime.datetime.fromtimestamp(df_2008['DepTime'].dropna().to_list())

TypeError: an integer is required (got type list)

In [ ]:
datetime.datetime.fromtimestamp(754.0)

In [41]:
df_2008[['DepTime','ArrTime']].dropna()

DepTime  ArrTime
0         2003.0   2211.0
1          754.0   1002.0
2          628.0    804.0
3          926.0   1054.0
4         1829.0   1959.0
...          ...      ...
7009723   1002.0   1204.0
7009724    834.0   1021.0
7009725    655.0    856.0
7009726   1251.0   1446.0
7009727   1110.0   1413.0

[6858079 rows x 2 columns]

### What is the structure of your dataset?

> There are 7,009,728 records of flight data in 2008 dataset recorded to show 29 different unique features of an on-time performance. Most variables are numeric time and time related variables recorded as floats should me converted to datetime type. Since the total size of all the files is 1.5 GB, despite trying to merge them under a single file was not successful. At this time it is not possible to do a time series analysis.

### What is/are the main feature(s) of interest in your dataset?

> I am most interested in figuring out if the delays are connected to a specific origin or destination also if there would be any attachment to a carrier or delays occur on a specific day of the week.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> DayOfWeek, DepTime, ArrTime, UniqueCarrier, AirTime, ArrDelay, DepDelay ,Distance, Origin and Dest columns would be the most important data for me to bring my analyses together. 

## Univariate Exploration

> Starting with the data type conversion then working with the dayofweek variable to see if there is any special day for the delay to occur

In [112]:
# df_2008['DepTime'] = pd.to_datetime(df_2008['DepTime'], format="%H:%M") 
# df_2008['DepTime']

#df_2008['Test'] = df_2008['DepTime'].map(lambda date_string: datetime.strptime(date_string, format_string))
df_2008['DepTime'].astype(str)


df_2008['DepTime'] = [str(x)[:-4]+':'+str(x)[-4:-2]+":"+str(x)[-2:] for x in df_2008['DepTime'].astype('int32')*100]
df_2008['DepTime'] = pd.to_datetime(df_2008['DepTime']).dt.strftime('%H:%M:%S')

#df_2008['DepTime'] = pd.to_datetime( df_2008['DepTime'])

ValueError: Cannot convert non-finite values (NA or inf) to integer

In [111]:
test = df_2008.head()


test['DepTime'] = [str(x)[:-4]+':'+str(x)[-4:-2]+":"+str(x)[-2:] for x in test['DepTime'].astype('int32')*100]
test['DepTime'] = pd.to_datetime(test['DepTime']).dt.strftime('%H:%M:%S')

test.head()


<ipython-input-111-73686aa05a1c>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test['DepTime'] = [str(x)[:-4]+':'+str(x)[-4:-2]+":"+str(x)[-2:] for x in test['DepTime'].astype('int32')*100]
<ipython-input-111-73686aa05a1c>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test['DepTime'] = pd.to_datetime(test['DepTime']).dt.strftime('%H:%M:%S')


Year  Month  DayofMonth  DayOfWeek   DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4  20:03:00        1955   2211.0   
1  2008      1           3          4  07:54:00         735   1002.0   
2  2008      1           3          4  06:28:00         620    804.0   
3  2008      1           3          4  09:26:00         930   1054.0   
4  2008      1           3          4  18:29:00        1755   1959.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   
3        1100            WN       1746  ...    3.0      7.0          0   
4        1925            WN       3920  ...    3.0     10.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           NaN          NaN      NaN   
1               NaN         0           NaN          NaN      NaN   
2               NaN         0           NaN          NaN      NaN   
3               NaN         0           NaN          NaN      NaN   
4               NaN         0           2.0          0.0      0.0   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            0.0               32.0  

[5 rows x 29 columns]

In [60]:


pd.to_datetime(df_2008['DepTime']).dt.strftime('%H:%M:%S')


0          00:00:00
1          00:00:00
2          00:00:00
3          00:00:00
4          00:00:00
             ...   
7009723    00:00:00
7009724    00:00:00
7009725    00:00:00
7009726    00:00:00
7009727    00:00:00
Name: DepTime, Length: 7009728, dtype: object

In [64]:
df_2008.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   2003.0        1955   2211.0   
1  2008      1           3          4    754.0         735   1002.0   
2  2008      1           3          4    628.0         620    804.0   
3  2008      1           3          4    926.0         930   1054.0   
4  2008      1           3          4   1829.0        1755   1959.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   
3        1100            WN       1746  ...    3.0      7.0          0   
4        1925            WN       3920  ...    3.0     10.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           NaN          NaN      NaN   
1               NaN         0           NaN          NaN      NaN   
2               NaN         0           NaN          NaN      NaN   
3               NaN         0           NaN          NaN      NaN   
4               NaN         0           2.0          0.0      0.0   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            0.0               32.0  

[5 rows x 29 columns]

In [ ]:
#df_2008['Test'] = datetime.datetime.fromtimestamp(df_2008['DepTime'])


for i in df_2008['DepTime']:
    #df_2008['Test'] = datetime.datetime.fromtimestamp(row)
    #df_2008['Test'].iat[i]=datetime.datetime.fromtimestamp(row['DepTime'])
    print(datetime.datetime.fromtimestamp(i))
    

In [26]:
df_2008['Test'] = None

In [24]:
df_2008['DepTime'].dropna().apply(lambda x : datetime.datetime.fromtimestamp(x))

0         1969-12-31 19:33:23
1         1969-12-31 19:12:34
2         1969-12-31 19:10:28
3         1969-12-31 19:15:26
4         1969-12-31 19:30:29
                  ...        
7009723   1969-12-31 19:16:42
7009724   1969-12-31 19:13:54
7009725   1969-12-31 19:10:55
7009726   1969-12-31 19:20:51
7009727   1969-12-31 19:18:30
Name: DepTime, Length: 6873482, dtype: datetime64[ns]

In [25]:
pd.to_datetime(df_2008['DepTime'].dropna().astype(str).str[:-2])

#df_2008['DepTime'].astype({'timestamp':'datetime64[ns]'})

ValueError: Given date string not likely a datetime.

In [ ]:
pd.to_datetime(df_2008['DepTime'], format='%H:%M') 

In [ ]:
df_2008.head()

In [ ]:
binsize = 500
bins = np.arange(0, df_2008['DepTime'].max()+binsize, binsize)

plt.figure(figsize=[8, 5])
plt.hist(data = df_2008, x = 'DepTime', bins = bins)
plt.xlabel('Departure Time')
plt.show()

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!